sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/EliminateJoinByCombineAggregate.scala

```diff
@@ -0,0 +1,196 @@
+    ne match {
+      case Alias(ae: AggregateExpression, _) => ae.filter.isEmpty
+      case ae: AggregateExpression => ae.filter.isEmpty
+    }
+  }
+
+  // Merge the multiple Aggregates.
+  private def mergePlan(
+      left: LogicalPlan,
+      right: LogicalPlan): Option[(LogicalPlan, Map[Expression, Attribute], Seq[Expression])] = {
+    (left, right) match {
+      case (la: Aggregate, ra: Aggregate) =>
+        val mergedChildPlan = mergePlan(la.child, ra.child)
+        mergedChildPlan.map { case (newChild, outputMap, filters) =>
+          val rightAggregateExprs = ra.aggregateExpressions.map { ne =>
+            ne.transform {
+              case attr: Attribute =>
+                outputMap.getOrElse(attr.canonicalized, attr)
+            }.asInstanceOf[NamedExpression]
+          }
+
+          val mergedAggregateExprs = if (filters.length == 2) {
+            la.aggregateExpressions.map { ne =>
+              ne.transform {
+                case ae @ AggregateExpression(_, _, _, None, _) =>
+                  ae.copy(filter = Some(filters.head))
+              }.asInstanceOf[NamedExpression]
+            } ++ rightAggregateExprs.map { ne =>
+              ne.transform {
+                case ae @ AggregateExpression(_, _, _, None, _) =>
+                  ae.copy(filter = Some(filters.last))
+              }.asInstanceOf[NamedExpression]
+            }
+          } else {
+            la.aggregateExpressions ++ rightAggregateExprs
+          }
+
+          (Aggregate(Seq.empty, mergedAggregateExprs, newChild), Map.empty, Seq.empty)
+        }
+      case (lp: Project, rp: Project) =>
+        val mergedInfo = mergePlan(lp.child, rp.child)
+        val mergedProjectList = ArrayBuffer[NamedExpression](lp.projectList: _*)
+
+        mergedInfo.map { case (newChild, outputMap, filters) =>
+          val allFilterReferences = filters.flatMap(_.references)
+          val newOutputMap = (rp.projectList ++ allFilterReferences).map { ne =>
+            val mapped = ne.transform {
+              case attr: Attribute =>
+                outputMap.getOrElse(attr.canonicalized, attr)
+            }.asInstanceOf[NamedExpression]
+
+            val withoutAlias = mapped match {
+              case Alias(child, _) => child
+              case e => e
+            }
+
+            val outputAttr = mergedProjectList.find {
+              case Alias(child, _) => child semanticEquals withoutAlias
+              case e => e semanticEquals withoutAlias
+            }.getOrElse {
+              mergedProjectList += mapped
+              mapped
+            }.toAttribute
+            ne.toAttribute.canonicalized -> outputAttr
+          }.toMap
+
+          (Project(mergedProjectList.toSeq, newChild), newOutputMap, filters)
+        }
+      // TODO support only one side contains Filter
+      case (lf: Filter, rf: Filter) =>
+        val mergedInfo = mergePlan(lf.child, rf.child)
+        mergedInfo.map { case (newChild, outputMap, _) =>
```

Comment:
I'm not sure this is correct as you simply ignore the propagated filters to here, but I can imagine adjacent `Filter` nodes in a plan...
Please check https://github.com/apache/spark/pull/37630 how it handles this case, but as I mentioned IMO the plan merge code shouldn't be duplicated to here.

Reply:
Currently, this PR support some of the most basic forms do not need to propagate filters here temporarily.

Reply:
I mean this case branch seems to behave incorrectly when you have adjacent filters in both plans.
E.g. if you have:
```
SELECT avg(a)
FROM (
  SELECT * FROM t WHERE c1 = 1
) t
WHERE c2 = 1
```
and
```
SELECT sum(b)
FROM (
  SELECT * FROM t WHERE c1 = 2
) t
WHERE c2 = 2
```
and `PushDownPredicates` is disabled then you are in that situation. When you are merging the parent `Filter` nodes you shouldn`t ignore the propagated filters from child `Filter` nodes.

Actually https://github.com/apache/spark/pull/37630 handles that case and many others and that's why I think splitting this PR into 2 parts would be better:
- This rule would be still useful if you used the existing `MergeScalarSubqueries.tryMergePlans()` to merge plans. It couldn't deal with Q28 immediately, but could deal with simpler aggregates without differing filters.
- https://github.com/apache/spark/pull/37630 could be a good base for implementing the filter merging feature. We could simplify that PR and remove the tricky physical plan checking part and introduce a simple boolean flag to enable/disable the filter merging feature like this PR does. In this case both Q9 and Q28 could benefit from the new feature...

Reply:
I think the user can disable `PushDownPredicates` with `spark.sql.optimizer.excludedRules` and `spark.sql.adaptive.optimizer.excludedRules` configs. And there might be other cases when we end up in that situation, that was just and example...

Reply:
Yeah, if users arbitrarily exclude the optimizer, then I believe most rules will execute improperly.

Reply:
No, otimizer rules might not make plans any better if prerequisites aren't met, but a rule shouldn't create incorrect plan in any circumstances.

Reply:
I got you mean now. Let's fix it.